lib/route/link/bridge.c

```diff
@@ -915,6 +1090,138 @@ uint16_t rtnl_link_bridge_str2hwmode(const char *name)
+ * Set port vlan membership range
+ * @arg link		Link object of type bridge
+ * @arg start		Start of membership range.
+ * @arg end			End of membership range.
+ * @arg untagged	Set membership range to be untagged.
+ *
+ * This will set the vlan membership range for a bridge port.
+ * Supported range is 0-4095
+ *
+ * @return 0 on success or negative error code
+ * @return -NLE_NOATTR		if port vlan attribute not present
+ * @return -NLE_OPNOTSUP 	Link is not a bridge
+ * @return -NLE_INVAL 		range is not in supported range.
+ */
+int rtnl_link_bridge_set_port_vlan_map_range (struct rtnl_link *link, uint16_t start, uint16_t end, int untagged)
+{
+	struct rtnl_link_bridge_vlan * vinfo;
+
+	IS_BRIDGE_LINK_ASSERT(link);
+
+	vinfo = rtnl_link_bridge_get_port_vlan(link);
+
+	if (!vinfo)
+		return -NLE_NOATTR;
+
+	if (start > end || end >= RTNL_LINK_BRIDGE_VLAN_BITMAP_MAX)
+		return NLE_INVAL;
+
+	for (uint16_t i = start; i <= end; i++)
+	{
+		set_bit(i,vinfo->vlan_bitmap);
+		if (untagged)
+			set_bit(i,vinfo->untagged_bitmap);
+	}
+	return 0;
+}
+
+/**
+ * Unset port vlan membership range
+ * @arg link		Link object of type bridge
+ * @arg start		Start of membership range.
+ * @arg end			End of membership range.
+ *
+ * This will unset the vlan membership range for a bridge port.
+ * Supported range is 0-4095
+ *
+ * @return 0 on success or negative error code
+ * @return -NLE_NOATTR		if port vlan attribute not present
+ * @return -NLE_OPNOTSUP 	Link is not a bridge
+ * @return -NLE_INVAL 		range is not in supported range.
+ */
+int rtnl_link_bridge_unset_port_vlan_map_range (struct rtnl_link *link, uint16_t start, uint16_t end)
+{
+	struct rtnl_link_bridge_vlan * vinfo;
+
+	IS_BRIDGE_LINK_ASSERT(link);
+
+	vinfo = rtnl_link_bridge_get_port_vlan(link);
+
+	if (!vinfo)
+		return -NLE_NOATTR;
+
+	if (start > end || end >= RTNL_LINK_BRIDGE_VLAN_BITMAP_MAX)
+		return -NLE_INVAL;
+
+	for (uint16_t i = start; i <= end; i++)
+	{
+		if (i != vinfo->pvid)
+		{
+			unset_bit(i,vinfo->vlan_bitmap);
+			unset_bit(i,vinfo->untagged_bitmap);
+		}
+	}
+	return 1;
+}
+
+/**
+ * Set port primary vlan id
+ * @arg link		Link object of type bridge
+ * @arg pvid		PVID to set.
+ * @arg untagged	Set vlan id to be untagged.
+ *
+ * This will set the primary vlan id for a bridge port.
+ * Supported range is 0-4095
+ *
+ * @return 0 on success or negative error code
+ * @return -NLE_NOATTR		if port vlan attribute not present
+ * @return -NLE_OPNOTSUP 	Link is not a bridge
+ * @return -NLE_INVAL 		PVID is above supported range.
+ */
+int rtnl_link_bridge_set_port_vlan_pvid (struct rtnl_link *link, uint16_t pvid, int untagged)
+{
+	struct rtnl_link_bridge_vlan * vinfo;
+
+	IS_BRIDGE_LINK_ASSERT(link);
+
+	vinfo = rtnl_link_bridge_get_port_vlan(link);
+
+	if (!vinfo)
+		return -NLE_NOATTR;
+
+	if (pvid >= RTNL_LINK_BRIDGE_VLAN_BITMAP_MAX)
+		return -NLE_INVAL;
+
+	vinfo->pvid = pvid;
+
+	set_bit(pvid,vinfo->vlan_bitmap);
+	if (untagged)
+		set_bit(pvid,vinfo->untagged_bitmap);
```

Comment:
setting the pvid sets also these bits.

However, when another pvid was set previously, it does not unset it.

I think, `rtnl_link_bridge_set_port_vlan_map_range()`, `rtnl_link_bridge_unset_port_vlan_map_range()` and `rtnl_link_bridge_set_port_vlan_pvid()` should not have this odd interaction. Which is also inconsistent. For example, if you set `rtnl_link_bridge_set_port_vlan_map_range()` with `untagged=false`, then `rtnl_link_bridge_set_port_vlan_map_range(tagged=true)` will overwrite it.

I think these functions should be completely separate. Add in the documentation of `rtnl_link_bridge_set_port_vlan_pvid()`, that the user most likely also want to call `rtnl_link_bridge_set_port_vlan_map_range()` for that pvid.